### ISSUE TITLE
how to use llama2 for instruction?

### ISSUE THREAD

**Issue by moseshu** [facebookresearch/llama#435]
Can i use llama2 to train alpaca,vicuna,orca  instruction with lora?do I need to change another prompt to do instruction?which is better?

**Comment by iibw** [comment]
I have the same question. I've been doing some digging and I think I have found the way.

**tl;dr**
```
<s>[INST] {user_instruction} [/INST] {model_response}</s>
```
---
First, according to the [README](https://github.com/facebookresearch/llama#fine-tuned-chat-models) it seems like there are two prompt formats that Llama was trained for. One for instruction and one for chat.

We know the one for chat because it is mentioned in the link above and can be reverse engineered from the referenced code. It is simply:
```
<s>[INST] <<SYS>>
{your_system_message}
<</SYS>>

{user_message_1} [/INST] {model_reply_1}</s><s>[INST] {user_message_2} [/INST] 
```
and the pattern continues for each turn in the conversation.

Instruction was a bit harder to find, but I think I figured it out. Instructions are defined as you tell the model to do something and then it does it. There is no system prompt or turn by turn chat history. Only the instruction you provide and the response of the model to that instruction.

Well, the prompt above without the system part or the chat history would be:
```
<s>[INST] 

{user_instruction} [/INST] {model_response}</s>
```
and then I found [a comment in the transformers codebase](https://github.com/huggingface/transformers/blob/9ef5256dfb28f5419649af8ea94e82573a6b2e77/src/transformers/models/llama/tokenization_llama.py#L340C9-L340C9) made by someone from HF who was working with Meta and probably knows the proper format of Llama 2 and it indicated that the instruction prompt was exactly what we thought: the chat prompt format without the system and turn by turn components. The only difference is the line breaks, but maybe those only exist when you include the system prompt and they should be removed.

This gives us the final instruction prompt of:
```
<s>[INST] {user_instruction} [/INST] {model_response}</s>
```

I spent way too long on this. Hope it helps!

**Comment by moseshu** [comment]
> 

what is your your_system_message for instruction prompt? 
it's like that <s>[INST] follow the below instruction give the answer\n my data..  [/INST]{response}</s>?is it right?
I have a problem that how to train an incremental data with lora, and retaining the original capabilities of the model.
now I always train the whole data (original data and incremental data) with lora. it's slow.

**Comment by iibw** [comment]
I don't have much experience with training LLMs, but I think you have the prompt right.

For example, if I wanted to translate some text, you'd do:
`<s>[INST] Translate the English text "Hello! How are you?" to Spanish. [/INST] {model_response}</s>`

**Comment by moseshu** [comment]
> 

many thanks  for your reply

**Comment by moseshu** [comment]
I try different prompts like 
```
def generate_prompt(content:str):
    header = (
              "A chat between a curious human and an artificial intelligence assistant. "
              "The assistant gives helpful, detailed, and polite answers to the human's questions.\n\n"
          )

    prompt2 = "[INST] You are a helpful, respectful and honest assistant.\n\n{instruction}\n[/INST]"

    prompt = "follow the below instruction to give the answer."
    start_msg = "### Human: {prompt}\n\n{instruction}.\n### Assistant: "
    prompt_result = start_msg.format_map({"prompt":prompt,"instruction":content})
    res = header +  prompt_result 
    return res

```
it works well,I use the above prompt get good result. Also ,you can change the header to any prompt like prompt2. alpaca,vicuna and so on. it does't have much effect on the result
